---
title: "Final CHARR figures"
output: html_document
date: "2023-06-19"
---

# Data Setup
```{r, echo=FALSE, message=FALSE, results='hide', warning=FALSE}
source('~/CHARR/R/constants.R')
```

## gnomAD v3 - WGS
### Load Data
```{r, message=FALSE, results='hide', warning=FALSE}
v3_meta <- read_tsv(paste0(data_path,'gnomad_v3_meta.tsv')) %>%
  filter(release & s != 'LP6005441-DNA_A09') %>%
  filter(label == 'broad' | hgdp)
v3_dp20 <- read_tsv(paste0(data_path,'gnomad_v3_contamination_estimate_full_snp_bi_minimum_vds_af_adjust_dp20_100.tsv')) %>%
  filter(s %in% v3_meta$s) 
v3_dp10 <- read_tsv(paste0(data_path,'gnomad_v3_contamination_estimate_full_snp_bi_minimum_vds_af_adjust_dp10_100.tsv')) %>%
  filter(s %in% v3_meta$s)
v3_no_dp <- read_tsv(paste0(data_path,'gnomad_v3_contamination_estimate_full_snp_bi_minimum_vds_af_adjust_no_dp.tsv')) %>%
  filter(s %in% v3_meta$s)
v3_dp20_indel <- read_tsv(paste0(data_path,'gnomad_v3_contamination_estimate_full_indel_bi_minimum_vds_af_adjust_dp20.tsv')) %>%
  filter(s %in% v3_meta$s)
v3_dp20_all <- read_tsv(paste0(data_path,'gnomad_v3_contamination_estimate_full_all_bi_minimum_vds_af_adjust_dp20.tsv')) %>%
  filter(s %in% v3_meta$s)
```

```{r}
## PCR_info
v3_pcr <- read.csv(paste0(data_path, 'rev_data/gnomad_v3_bimodality_check.csv'), sep='\t')
hgdp_info <- read_tsv(paste0(data_path,'HGDP_meta_v1.tsv')) %>% 
  filter(subsets.hgdp) %>% 
  select(s, pcr = bergstrom.library_type, source=bergstrom.source) 
v3_pcr <- v3_pcr %>%
  merge(., hgdp_info, by = 's', all.x = T) %>%
  dplyr::mutate(pcr_free = if_else(is.na(pcr_free), pcr, pcr_free)) %>%
  mutate(pcr_free = if_else(is.na(pcr_free), 'Missing', pcr_free) )%>%
  mutate(pcr_free = if_else(pcr_free %in% c('No', 'PCR'), 'PCR', pcr_free)) %>%
  mutate(pcr_free = if_else(pcr_free %in% c('PCR-free', 'PCRfree'), 'PCR-Free', pcr_free)) %>%
  select(s, pcr_free)
v3_dp20 <- v3_dp20 %>%
  merge(., v3_pcr, by = 's')
```

### Convert to long format
```{r, message=FALSE, results='hide', warning=FALSE}
v3_dp20_long <- convert_to_long_format(v3_dp20, 'v3')
v3_dp10_long <- convert_to_long_format(v3_dp10, 'v3')
v3_no_dp_long <- convert_to_long_format(v3_no_dp, 'v3')
v3_dp20_long_indel <- convert_to_long_format(v3_dp20_indel, 'v3')
v3_dp20_long_all <- convert_to_long_format(v3_dp20_all, 'v3')
```

## gnomAD v2 - WES
### Load Data
```{r, message=FALSE, results='hide', warning=FALSE}
v2_meta <- read_tsv(paste0(data_path,'gnomad_v2_meta.tsv')) %>%
  filter(v2_release & project != 'C774' & n_het > 10000) 
v2_dp20 <- read_tsv(paste0(data_path,'gnomad_v2_contamination_estimate_full_snp_bi_minimum_vds_af_adjust_dp20_100.tsv')) %>%
  filter(s %in% v2_meta$s)
v2_dp10 <- read_tsv(paste0(data_path,'gnomad_v2_contamination_estimate_full_snp_bi_minimum_vds_af_adjust_dp10_100.tsv')) %>%
  filter(s %in% v2_meta$s)
v2_no_dp <- read_tsv(paste0(data_path,'gnomad_v2_contamination_estimate_full_snp_bi_minimum_vds_af_adjust_no_dp.tsv')) %>%
  filter(s %in% v2_meta$s)
```

### Convert to long format
```{r, message=FALSE, results='hide', warning=FALSE}
v2_dp20_long <- convert_to_long_format(v2_dp20, 'v2')
v2_dp10_long <- convert_to_long_format(v2_dp10, 'v2')
v2_no_dp_long <- convert_to_long_format(v2_no_dp, 'v2')
```

# Main Figure 2
```{r}
v3_main <- v3_dp20 %>% 
  mutate(label = factor(label, labels = c('Broad', 'Other'), levels = c('broad', 'hgdp')),
         contamination = if_else(!is.na(recomputed_contamination), recomputed_contamination, contamination),
         data_type = '(A): gnomAD v3 (WGS)') %>%
  select(mean_AB_snp_biallelic_af_adjust_10, n_snp_biallelic_af_10, contamination, label, data_type)
v2_main <- v2_dp20 %>% 
  select(mean_AB_snp_biallelic_af_adjust_10, n_snp_biallelic_af_10, contamination, label) %>%
  mutate(label = if_else(label, 'Broad', 'Other'),
         data_type = '(B): gnomAD v2 (WES)') 
main_data <- rbind(v3_main %>% filter(label == 'Broad'), v3_main %>% filter(label == 'Other'), v2_main)
main_var_cnt <-   main_data %>%
    group_by(data_type) %>%
    dplyr::summarize(mean_n_var = mean(n_snp_biallelic_af_10))
```

```{r}
library(dplyr)
main_data %>%
    group_by(label, data_type) %>% 
    dplyr::summarize(tidy(cor.test(contamination, mean_AB_snp_biallelic_af_adjust_10)))
main_data %>%
    group_by(data_type) %>% 
    dplyr::summarize(tidy(cor.test(contamination, mean_AB_snp_biallelic_af_adjust_10)))
```

```{r}
figure <- main_data %>%
  ggplot + aes(x = contamination, y = mean_AB_snp_biallelic_af_adjust_10, color = label) +
  labs(x = 'Freemix Score', y = paste0('CHARR Score'), color = 'Sequencing site') +
  geom_point(alpha=0.8, size = 1) + 
  geom_abline(slope = 1, intercept = 0, lty=2) +
  scale_color_manual(values = c('#045494', 'gray')) +
  themes +  theme_classic() + 
  theme(axis.title = element_text(family = 'Arial', size = 15, face = 'bold'),
        legend.text = element_text(family = 'Arial', size = 12),
        legend.title = element_text(family = 'Arial', size = 12, face = 'bold'),
        legend.position = 'top',
        axis.text = element_text(family = 'Arial', size = 12),
        strip.text.x = element_text(family = 'Arial', size = 10, face = 'bold')) + 
  facet_wrap(.~data_type, scale = 'free') 
# +
#   geom_label(data = main_var_cnt, aes(label = paste('Mean(N_hom_var) = ', as.character(round(mean_n_var)))),
#              x = 0.013, y = Inf, vjust = 1,  color = 'black', size = 3, family = 'Arial') +
#   annotate(geom='text', x= 0.048, y=Inf,vjust = 1, label='y=x', family = 'Arial', size=6)
save <- T
if(save){
  png(paste0(figure_path, 'M2_gnomad_contam_dp20_ref10_no_label.png'), height = 4, width = 7.5, units = 'in', res = 300)
  print(figure)
  dev.off()
}
figure
```

```{r}
figure <- main_data %>%
  select(estimator = mean_AB_snp_biallelic_af_adjust_10, label, data_type) %>% 
  mutate(score_type='CHARR Score') %>%
  rbind(main_data %>% select(estimator = contamination, label, data_type) %>% 
  mutate(score_type='Freemix Score')) %>%
  ggplot + aes(x = estimator, color = label) +
  labs(x = 'Contamination Estimator', y='Density', color = 'Sequencing site') +
  geom_density() +
  # geom_point(alpha=0.8, size = 1) + 
  # geom_abline(slope = 1, intercept = 0, lty=2) +
  scale_color_manual(values = c('#045494', 'gray')) +
  # geom_vline(data = data.frame(data_type = c('(A): gnomAD v3 (WGS)', '(B): gnomAD v2 (WES)'), ref_line = c(0.003, 0.01), ), aes(xintercept = ref_line), lty=2) +
  themes +  theme_classic() + 
  theme(axis.title = element_text(family = 'Arial', size = 15, face = 'bold'),
        legend.text = element_text(family = 'Arial', size = 12),
        legend.title = element_text(family = 'Arial', size = 12, face = 'bold'),
        legend.position = 'top',
        axis.text = element_text(family = 'Arial', size = 12),
        strip.text = element_text(family = 'Arial', size = 10, face = 'bold')) + 
  scale_x_log10() + 
  facet_grid(score_type~data_type, scale = 'free') 
# +
#   geom_label(data = main_var_cnt, aes(label = paste('Mean(N_hom_var) = ', as.character(round(mean_n_var)))),
#              x = 0.013, y = Inf, vjust = 1,  color = 'black', size = 3, family = 'Arial') +
#   annotate(geom='text', x= 0.048, y=Inf,vjust = 1, label='y=x', family = 'Arial', size=6)
save <- T
if(save){
  png(paste0(figure_path, 'SX_log_density_gnomad_contam_dp20_ref10_no_label.png'), height = 6, width = 7.5, units = 'in', res = 300)
  print(figure)
  dev.off()
}
figure
```

# Main Figure 3
```{r}
save <- T
fig3_n_way_comparison('M3', save=save, height = 4, width = 6.5)
```

# Supplement Figure 1 | Number of high-quality homozygous variants with RR > 0
```{r}
save <- T
figS1_n_hom_var_RR_above_0(save=save, height = 4, width = 6)
```

# Table.S1 | Number of high-quality homozygous variants with RR > X & refAB > 10%
```{r}
hom_var <- read_tsv(paste0(data_path, 'gnomad_v3_n_hom_var_final.tsv'))%>%
  mutate(label = if_else(hgdp, 'hgdp', label),
        contamination = if_else(label == 'broad', contamination/100, contamination),) %>%
      mutate(label = factor(label, labels = c('Broad (gnomAD)', 'Sanger (HGDP)'), levels = c('broad', 'hgdp')),
             contamination = if_else(!is.na(recomputed_contamination), recomputed_contamination, contamination),) 
```

```{r}
sum(hom_var$n_hom_var_RR_above_0 > 0)
sum(hom_var$n_hom_var_RR_above_0 > 10000)
sum(hom_var$n_hom_var_RR_above_0 > 20000)
sum(hom_var$n_hom_var_RR_above_0 > 30000)
sum(hom_var$n_hom_var_RR_above_0 > 100000)
sum(hom_var$n_hom_var_RR_above_0 > 200000)
```

```{r}
sum(hom_var$n_hom_var_refAB_above_10pct > 0)
sum(hom_var$n_hom_var_refAB_above_10pct > 30)
sum(hom_var$n_hom_var_refAB_above_10pct > 50)
sum(hom_var$n_hom_var_refAB_above_10pct > 100)
sum(hom_var$n_hom_var_refAB_above_10pct > 500)
sum(hom_var$n_hom_var_refAB_above_10pct > 1000)
```

# Supplement Figure 2 | gnomAD v3 with and without ref AF filter 
```{r}
save <- T
get_figure_charr_freemix_by_ref_AF(v3_dp20, version = 'v3', DP = '20', var_type =  'snp', save = save, filter = TRUE, sub_ref_AF = c('ref-AF: (0%, 100%)', 'ref-AF: (10%, 90%)'), height=4, width=7.5 )
```

# Supplement Figure 3 | gnomAD v3 Linear regression with different DP filters
```{r, fig.width=4cm}
intercept <- T
save <- T
s3a <- figS3_get_figure_lm_by_dp_ref_AF(long_data_no_dp = v3_no_dp_long %>% mutate(label = factor(label, levels = c('Broad (gnomAD)', 'Sanger (HGDP)'), labels = c('Broad', 'Other'))), 
                                long_data_dp10 =  v3_dp10_long %>% mutate(label = factor(label, levels = c('Broad (gnomAD)', 'Sanger (HGDP)'), labels = c('Broad', 'Other'))),
                                long_data_dp20 =  v3_dp20_long %>% mutate(label = factor(label, levels = c('Broad (gnomAD)', 'Sanger (HGDP)'), labels = c('Broad', 'Other'))), 
                                 version='v3', var_type = 'snp', 
                                 name='S3A', intercept, save)
s3b <- figS3_get_figure_lm_by_dp_ref_AF(long_data_no_dp = v2_no_dp_long, long_data_dp10 = v2_dp10_long, long_data_dp20 = v2_dp20_long, version='v2', var_type = 'snp', name='S3B', intercept, save)
```

```{r}
save <- T
figure = ggpubr::ggarrange(s3a, NULL, s3b, labels = c('(A) gnomAD v3 (WGS)', '', '(B) gnomAD v2 (WES)'), nrow=3, common.legend=TRUE, vjust = 0, hjust = 0, font.label = list(size = 10, color = "black", face = "bold", family = NULL),heights = c(0.18, 0.02, 0.18))
figure = annotate_figure(figure, top = text_grob(paste0('CHARR score ~ Freemix score (SNP)'), face = 'bold', size = 15))
png(paste0(figure_path,'S3_gnomad_contam_dp_all_snp_linear_model.png'), height = 10, width = 10, units='in', res=300)
print(figure)
dev.off()
```

# Supplement Figure 4 | gnomAD v3 CHARR vs. Freemix Score for SNPs INDELS and ALL linear model
```{r}
save <- T
v3_full <- rbind(v3_dp20 %>% select(-pop), v3_dp20_indel, v3_dp20_all) %>% mutate(var_type = rep(c('SNPs', 'INDELs', 'ALL'), each = 59765))
v3_long_full <- convert_to_long_format(v3_full, 'v3')
figS4_v3_snp_indel_all_lm(v3_long_full, 'v3', 'S4', save, height = 6, width = 10)
```

# Supplement Figure 5 | gnomAD v2 different reference allele frequency sources
```{r}
v2_dp20_new <- read_tsv(paste0(data_path,'gnomad_v2_contamination_estimate_full_snp_bi_minimum_vds_using_v3_af_adjust_dp20_100.tsv')) %>%
  filter(s %in% v2_meta$s)
```

```{r}
v2_dp20_long_data_new <- convert_to_long_format(v2_dp20_new, 'v2')  %>% mutate(AF_source = 'AF Source: v3 (WGS)')
v2_dp20_var_cnt_data_new <- get_n_hom_var_table(v2_dp20_new  %>% mutate(AF_source = 'AF Source: v3 (WGS)'))
v2_dp20_long_data <- convert_to_long_format(v2_dp20, 'v2') %>% mutate(AF_source = 'AF Source: v2 (WES)')
v2_dp20_var_cnt_data <- get_n_hom_var_table(v2_dp20 %>% mutate(AF_source = 'AF Source: v2 (WES)'))
```

```{r}
v2_dp20_long_data_full <- rbind(v2_dp20_long_data_new, v2_dp20_long_data) %>% filter(ref_AF %in% c('ref-AF: (1%, 99%)', 'ref-AF: (5%, 95%)', 'ref-AF: (10%, 90%)', 'ref-AF: (20%, 80%)'))
v2_dp20_var_cnt_data_full <- rbind(v2_dp20_var_cnt_data_new, v2_dp20_var_cnt_data) %>% filter(ref_AF %in% c('ref-AF: (1%, 99%)', 'ref-AF: (5%, 95%)', 'ref-AF: (10%, 90%)', 'ref-AF: (20%, 80%)'))
```

```{r}
library(dplyr)
v2_dp20_long_data_full  %>%
    group_by(ref_AF, AF_source) %>% 
    dplyr::summarize(tidy(cor.test(contamination, value)))
```

```{r}
save <- T
print(v2_dp20_var_cnt_data_full)
figS5_v2_ref_af_source(v2_dp20_long_data_full, v2_dp20_var_cnt_data_full, 'S5', save=save, height = 6, width = 9)
```


# Supplement Figure 6 | gnomAD v2 different reference allele frequency sources linear regression
```{r}
save <- T
figS6_v2_ref_af_source_lm(long_data = v2_dp20_long_data_full, 'S6', save=save, height = 6, width = 10)
```

# Supplement Figure 7 | gnomAD v3 with and without the 3 age genes:  ASXL1, TET2, DNMT3A
```{r, message=FALSE, results='hide', warning=FALSE}
v3_dp20_age <- read_tsv(paste0(data_path,'gnomad_v3_contamination_estimate_full_snp_bi_minimum_vds_af_adjust_dp20_age_gene.tsv')) %>%
  filter(s %in% v3_meta$s)
v3_dp20 <- read_tsv(paste0(data_path,'gnomad_v3_contamination_estimate_full_snp_bi_minimum_vds_af_adjust_dp20_100.tsv')) %>%
  filter(s %in% v3_meta$s)
```

```{r}
v3_dp20_long_data_age <- convert_to_long_format(v3_dp20_age, 'v3') %>% mutate(source = '3 age gene removed')
v3_dp20_var_cnt_data_age <- v3_dp20_age %>% filter(label != 'external') %>% get_n_hom_var_table(.) %>% mutate(source = '3 age gene removed')
v3_dp20_long_data <- convert_to_long_format(v3_dp20, 'v3') %>% mutate(source = 'All SNPs') %>% select(-pop)
v3_dp20_var_cnt_data <- v3_dp20 %>% filter(label != 'external') %>% get_n_hom_var_table(.) %>% mutate(source = 'All SNPs')
```

```{r}
v3_dp20_long_data_full <- rbind(v3_dp20_long_data_age, v3_dp20_long_data) %>% filter(ref_AF %in% c('ref-AF: (1%, 99%)', 'ref-AF: (5%, 95%)', 'ref-AF: (10%, 90%)', 'ref-AF: (20%, 80%)') & label != 'Other')
v3_dp20_var_cnt_data_full <- rbind(v3_dp20_var_cnt_data_age, v3_dp20_var_cnt_data) %>% filter(ref_AF %in% c('ref-AF: (1%, 99%)', 'ref-AF: (5%, 95%)', 'ref-AF: (10%, 90%)', 'ref-AF: (20%, 80%)'))
```

```{r}
save <- T
figS7_v3_remove_age_gene(v3_dp20_long_data_full, v3_dp20_var_cnt_data_full, 'S7', save=save, height = 6, width = 9)
```


# Supplement Figure 8 | gnomAD v3 with and without the 3 age genes:  ASXL1, TET2, DNMT3A, linear regression
```{r}
save <- T
figS8_v3_remove_age_gene_lm(v3_dp20_long_data_full, 'S8', save=save, height = 6, width = 10)
```

# Supplement Figure 9 & Table S2 | Age vs. CHARR
```{r}
save <- T
figS9_v2_v3_age_vs_charr(v2_dp20_long, v3_dp20_long, 'S9', save=save, height = 4, width = 7.5)
```



```{r}
v2_long_data <- v2_dp20_long %>%
  filter(ref_AF == 'ref-AF: (10%, 90%)') %>%
  mutate(age_bin = factor(get_age_bins(age), levels = age_bins),
         version = 'gnomAD v2 (WES)') %>%
  select(value, ref_AF, age_bin, version, age, label) 
print('v2 LM')
print(summary(lm(value ~ age, data = v2_long_data)))
v3_long_data <- v3_dp20_long %>%
  filter(ref_AF == 'ref-AF: (10%, 90%)') %>%
  mutate(age_bin = factor(get_age_bins(age), levels = age_bins),
         version = 'gnomAD v3 (WGS)') %>%
  select(value, ref_AF, age_bin, version, age, label, pcr_free)

print('v3 LM CHARR ~ Age')
print(summary(lm(value ~ age, data = v3_long_data )))
print('v3 LM CHARR ~ Age + PCR')
print(summary(lm(value ~ age + pcr_free, data = v3_long_data %>% filter(pcr_free != 'Missing'))))
print('v3 LM CHARR ~ Age')
print(summary(lm(value ~ age, data = v3_long_data %>% filter(pcr_free != 'Missing'))))
```

```{r}
pcr_data <- v3_long_data %>% filter((pcr_free != 'Missing') & !is.na(age))
# Fit linear model to regress out 'am' from 'hp'
model <- lm(value~pcr_free, data = pcr_data)

# Calculate residuals
residuals <- residuals(model)

# Create a new data frame with residuals and corresponding mpg values
residuals_data <- data.frame(age = pcr_data$age, residuals = residuals)

# Plot residuals as a function of mpg
ggplot(residuals_data, aes(x = age, y = residuals)) +
  # geom_point(color='#045494', alpha=0.5) +
  geom_raster(aes(fill = density)) +
  # geom_density_2d() +
  geom_smooth(method = "lm", se = FALSE, color='black') +
  # scale_y_log10() +
  labs(x = "Age",
       y = "Residuals (CHARR ~ PCR status)", 
       color = "Count") + themes
```
```{r}
pcr_data <- v3_dp20 %>% filter((pcr_free != 'Missing') & !is.na(age))
model <- lm(mean_AB_snp_biallelic_af_adjust_10~pcr_free, data = pcr_data)
residuals <- residuals(model)
residuals_data <- data.frame(age = pcr_data$age, residuals = residuals)
figure <- residuals_data %>% 
  mutate(age_bin  = factor(get_age_bins(age), levels = age_bins) )%>%
  ggplot + aes(x = age_bin, y = residuals, group=age_bin) +
  geom_boxplot() +
  geom_smooth(method = "lm", se = FALSE, color='black') +
  # scale_y_log10() +
  labs(x = "Age",
       y = "Residuals (CHARR ~ PCR status)", 
       color = "Count") + themes
figure
png(paste0(figure_path, 'S10_gnomad_contam_dp20_age_vs_charr_pcr_residual.png'), height = 4, width = 6, units = 'in', res = 300)
print(figure)
dev.off()
```


# Supplement Figure 11
```{r}
p<- v3_dp20 %>%
  mutate(pcr_free = factor(pcr_free, levels=c('PCR-Free', 'PCR', 'Missing'), labels=c('PCR-Free', 'PCR+', 'Missing'))) %>%
  ggplot + aes(x = mean_AB_snp_biallelic_af_adjust_10, color = pcr_free, fill=pcr_free) +
  labs(x = 'CHARR', y='Density', color='Sequencing Method', fill = 'Sequencing Method') +
  geom_density(alpha=0.5) +
  scale_x_log10(label=comma) +
  scale_color_brewer(palette = 'Dark2') + 
  scale_fill_brewer(palette = 'Dark2') + 
  themes
p
png(paste0(figure_path, 'S11_gnomad_v3_charr_bimodality_pcr_free_hist_all.png'), height = 4, width = 6, units = 'in', res = 300)
print(p)
dev.off()
```


# Supplement Figure 12 | Freemix score old vs. recomputed version
```{r}
save <- T
hgdp <- read_tsv(paste0(data_path,'gnomad_v3_contamination_estimate_full_snp_bi_minimum_vds_af_adjust_dp20_100.tsv')) %>% 
  mutate(label = factor(label, labels = c('Broad (gnomAD)', 'Sanger (HGDP)', 'Other'), levels = c('broad', 'hgdp', 'external'))) 
figS12_hgdp_freemix_before_after(hgdp, 'S12', save = save, height = 4, width = 6)
```

# Supplement Figure 13 | CHARR vs. Freemix Score before and after on HGDP samples
```{r}
save <- T
figS13_hgdp_charr_freemix_before_after(hgdp %>% filter(s %in% v3_meta$s), 'S13', save=save, height = 5, width = 7.5)
```

# Supplement Figure 14 | HGDP sample selection
```{r}
save = T
fig14_hgdp_sample_selection('S14', save=save, height = 4, width = 10)
```

# Supplement Figure 15 | CHARR vs. Freemix Score before and after decontamination on the selected HGDP samples
```{r}
save= T
figS15_sample_decontamination('S15', save=save, height = 4, width = 7.5)
```

# Supplement Figure 16 | 2 way freemixscore vs. CHARR color by the truth
```{r}
save = T
figure1 = figS16_2_way_color_by_truth('S16A', save=save, height = 4, width = 6, type = 'gnomAD')
figure2 = figS16_2_way_color_by_truth('S16B', save=save, height = 4, width = 6, type = 'sample')
figure = ggpubr::ggarrange(figure1, figure2, ncol = 2, labels = c('(A): gnomAD AF', '(B): local AF'), common.legend = TRUE, vjust = 0.5, hjust=0,
                             #vjust = 0, hjust = 0, 
                             font.label = list(size = 13, color = "black", face = "bold", family = 'Arial'))
figure

png(paste0(figure_path, 'S16_hgdp_2_way_mixed_150_samples.png'), height = 4, width = 8, units = 'in', res = 300)
print(figure)
dev.off()
```

# Supplement Figure 17 | het/hom ratios of the two-way mixing results
```{r}
save = T
figS17_two_way_mixing_het_hom_ratio('S17', save=save, height = 4, width = 6)
```

# Supplement Figure 18 | N-way mixing results CHARR. vs. Freemix Score
```{r}
save = T
figS18_n_way_mixing('S18', save=save, height = 6, width = 8)
```

# Supplement Figure 19 | N-way mixing boxplot
```{r}
save = T
figS19_n_way_box_plot('S19_gnomad_', save=save, height = 3.5, width = 7)
```

# Supplement Figure 20 | two-way vs. N-way mixing results
```{r}
save = T
figS20_2_way_vs_n_way('S20', save=save, height = 4.5, width = 9, type = 'gnomAD')
figS20_2_way_vs_n_way('S20', save=save, height = 4.5, width = 9, type = 'sample')
```